MyRangesImplementation/MyRangesImplementation/cartezian_product_range_adapter.hpp

```diff
@@ -0,0 +1,60 @@
+#pragma once
+#include"range_traits.hpp"
+#include"cartezian_product_iterator.hpp"
+#include<vector>
+#include<algorithm>
+
+namespace ranges {
+	namespace view {
+		template<typename... Ranges>
+		class cartezian_product_range_adapter {
+			static_assert(RangeTraits::are_all_ranges<Ranges...>::value,
+				"Error: all types in the given list of types must be Ranges");
+			std::tuple<Ranges...> m_ranges;
+
+		public:
+			using value_type = std::tuple<typename Ranges::value_type...>;
+			using iterator = ranges::internals::iterators::cartezian_product_iterator<typename Ranges::const_iterator...>;
+
+			cartezian_product_range_adapter(const Ranges&... ranges)
+				noexcept(noexcept(std::make_tuple(ranges...)))
+				:m_ranges(std::make_tuple(ranges...)) {}
+
+			cartezian_product_range_adapter(const cartezian_product_range_adapter<Ranges...>&) = default;
+			cartezian_product_range_adapter(cartezian_product_range_adapter<Ranges...>&&) = default;
+
+			auto begin() const {
```

Comment:
non-const versions are not possible?

Reply:
views on ranges are all const because in views you cant change anything, its a view on existing data.